src/main/java/com/gradle/GitMetadata.java

```diff
@@ -0,0 +1,54 @@
+package com.gradle;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import static com.gradle.Utils.execAndGetStdOut;
+
+class GitMetadata {
+
+    private final String[] args;
+    private final boolean isGitInstalled;
+    private final Supplier<Optional<String>> fromEnv;
+
+    private GitMetadata(GitMetadataBuilder metadataBuilder) {
+        this.args = metadataBuilder.args;
+        this.isGitInstalled = metadataBuilder.isGitInstalled;
+        this.fromEnv = metadataBuilder.fromEnv;
+    }
+
+    Optional<String> resolve() {
+        if (fromEnv != null) {
+            return fromEnv.get();
+        }
+        if (isGitInstalled && args != null) {
+            return execAndGetStdOut(args);
+        }
+        return Optional.empty();
+    }
+
+    static class GitMetadataBuilder {
```

Comment:
in the current state, this feels a bit more like a `Factory` to me :D 

I'm trying to think of the best way to structure this though. i haven't really fully thought it out but the current implementation seems like it could be simplified.

Reply:
yeah, i think a factory would make sense since it would remove the case of the `.build().resolve()` that is used in every case.

Reply:
How about the latest refactoring a bit less expressive but simple: a resolver that just resolves :))